### PR TITLE
[sfputil] Configure the debug loopback mode only on the relevant lanes of the logical port

### PR DESCRIPTION
**Pull request by xinyulin**
#### What I did
This update enhances the loopback command functionality to support configuring the loopback mode for specific logical ports, instead of applying the configuration to the entire physical port.

#### How I did it
By calculating the lane mask according to the port's configuration and state data, the loopback mode can now be accurately set for the specified port, ensuring that other unrelated logical ports are not affected during diagnostic operations.

#### How to verify it
Tested under Cisco8111 with Credo C1 cable.

The sfputil debug loopback command will now only affect the relevant lanes of the logical port.

Set **host input** loopback mode on logical port **Ethernet80**, the expected value of **page 0x13, byte 0xB7= 0x0F**
image

Set **media input** loopback mode on logical port **Ethernet84**, the expected value of **page 0x13, byte 0xB5= 0xF0**
image

#### New command output (if the output of a command-line utility has changed)

### CODE REVIEW

File: sfputil/main.py [hunk 1]
```diff
@@ -1991,11 +1991,44 @@ def loopback(port_name, loopback_mode):
         click.echo("{}: This functionality is not implemented".format(port_name))
         sys.exit(ERROR_NOT_IMPLEMENTED)
 
+    namespaces = multi_asic.get_front_end_namespaces()
+    for namespace in namespaces:
```
[the remark this answers is not in the document]
Comment (xinyulin) [reply sfputil/main.py]: @mihirpat1 Thanks for the example, I have modified it.

File: sfputil/main.py [hunk 2]
```diff
@@ -1991,11 +1991,44 @@ def loopback(port_name, loopback_mode):
...
+        config_db = ConfigDBConnector(use_unix_socket_path=True, namespace=namespace)
+        if config_db is not None:
+            config_db.connect()
+            subport = int(config_db.get(config_db.CONFIG_DB,
```
[the remark this answers is not in the document]
Comment (xinyulin) [reply sfputil/main.py]: @mihirpat1 Could we simply assign the subport to 1 if the subport is not defined in the CONFIG_DB (subport is None) or is mistakenly set to 0? 
`        if subport is None or subport == 0:
            subport = 1`

[the remark this answers is not in the document]
Comment (mihirpat1) [reply sfputil/main.py]: @xinyulin Yes, this should be fine as well

File: sfputil/main.py [hunk 3]
```diff
@@ -1991,11 +1991,48 @@ def loopback(port_name, loopback_mode):
...
         sys.exit(ERROR_NOT_IMPLEMENTED)
 
+    namespace = multi_asic.get_namespace_for_port(port_name)
+    config_db = ConfigDBConnector(use_unix_socket_path=True, namespace=namespace)
+    if config_db is not None:
+        config_db.connect()
+        subport = int(config_db.get(config_db.CONFIG_DB,
+                                    'PORT|{}'.format(port_name), 'subport'))
+
+        # If it is not defined (None) or if it is set to 0, assign a default value of 1
```
[the remark this answers is not in the document]
Comment (xinyulin) [reply sfputil/main.py]: @mihirpat1 Thank you for correcting the wording. I also added more **loopback_mode** options to allow users to individually clear each mode in the latest commit.
```
@click.argument('loopback_mode', required=True, default="none",
                type=click.Choice(["none", "host-side-input", "host-side-output",
                                   "media-side-input", "media-side-output",
                                   "host-side-input-none", "host-side-output-none",
                                   "host-side-input-none", "host-side-output-none"]))
```
I’ve kept `none` as a valid loopback_mode option. Do you think it’s useful for users who want to completely clear all loopback modes on the physical port at once? Or should we remove this `none` option and require users to disable the loopback mode individually?

File: sfputil/main.py [hunk 4]
```diff
@@ -1991,11 +1993,50 @@ def loopback(port_name, loopback_mode):
...
+                                           'TRANSCEIVER_INFO|{}'.format(port_name),
+                                           'host_lane_count'))
+        media_lane_count = int(state_db.get(state_db.STATE_DB,
+                                            'TRANSCEIVER_INFO|{}'.format(port_name),
+                                            'media_lane_count'))
+    else:
+        click.echo("Failed to connect to STATE_DB")
+        sys.exit(EXIT_FAIL)
+
+    if loopback_mode in ("host-side-input", "host-side-output",
```
Comment (mihirpat1) [inline sfputil/main.py:2024]: Can we replace this block with something similar to the below?
```
    lane_mask = 0
    if "host-side" in loopback_mode:
        lane_mask = ((1 << host_lane_count) - 1) << ((subport - 1) * host_lane_count)
    elif "media-side" in loopback_mode:
        lane_mask = ((1 << media_lane_count) - 1) << ((subport - 1) * media_lane_count)
```

Comment (xinyulin) [reply sfputil/main.py]: Thanks! It looks compact and clean!

File: sfputil/main.py [hunk 5]
```diff
@@ -1991,11 +1993,48 @@ def loopback(port_name, loopback_mode):
...
+        if subport is None or subport == 0:
+            subport = 1
+    else:
+        click.echo("Failed to connect to CONFIG_DB")
+        sys.exit(EXIT_FAIL)
+
+    state_db = SonicV2Connector(use_unix_socket_path=False, namespace=namespace)
+    if state_db is not None:
+        state_db.connect(state_db.STATE_DB)
+        host_lane_count = int(state_db.get(state_db.STATE_DB,
```
Comment (mihirpat1) [inline sfputil/main.py:2014]: @xinyulin Wondering if we should handle ValueError exception and return error to handle case wherein `subport`, `host_lane_count` or `media_lane_count` are not present in the DB.

File: sfputil/main.py [hunk 6]
```diff
@@ -1469,10 +1469,10 @@ def is_fw_switch_done(port_name):
                 status = -1 # Abnormal status.
             elif (ImageARunning == 1) and (ImageACommitted == 0):   # ImageA is running, but not committed.
                 click.echo("FW images switch successful : ImageA is running")
-                status = 1  # run_firmware is done. 
+                status = 1  # run_firmware is done.
             elif (ImageBRunning == 1) and (ImageBCommitted == 0):   # ImageB is running, but not committed.
                 click.echo("FW images switch successful : ImageB is running")
-                status = 1  # run_firmware is done.
```
[the remark this answers is not in the document]
Comment (xinyulin) [reply sfputil/main.py]: Oops! My editor automatically trimmed the trailing whitespace. Let me revert it and disable that feature.

File: sfputil/main.py (hunk 5 above)
Comment (xinyulin) [reply sfputil/main.py]: @mihirpat1 It is helpful for the user to know about the incorrect configuration. The newly added commit catches the `TypeError` when the variable is not present in the DB. 
```python
try:
    subport = int(config_db.get(config_db.CONFIG_DB, f'PORT|{port_name}', 'subport'))
except TypeError:
    click.echo(f"{port_name}: subport is not present in CONFIG_DB")
    sys.exit(EXIT_FAIL)
``` 
From the formula, the `lane_mask` will automatically be set to 0 when the host or media lane count is zero. 
```python
if 'host-side' in loopback_mode:
    lane_mask = ((1 << host_lane_count) - 1) << ((subport - 1) * host_lane_count)
elif 'media-side' in loopback_mode:
    lane_mask = ((1 << media_lane_count) - 1) << ((subport - 1) * media_lane_count)
```

File: sfputil/main.py [hunk 7]
```diff
@@ -1968,10 +1968,12 @@ def debug():
 # 'loopback' subcommand
 @debug.command()
 @click.argument('port_name', required=True, default=None)
-@click.argument('loopback_mode', required=True, default="none",
-                type=click.Choice(["none", "host-side-input", "host-side-output",
+@click.argument('loopback_mode', required=True, default=None,
+                type=click.Choice(["host-side-input", "host-side-output",
                                    "media-side-input", "media-side-output"]))
-def loopback(port_name, loopback_mode):
+@click.argument('enable', required=True, default=None,
```
Comment (mihirpat1) [inline sfputil/main.py:1974]: @xinyulin Why do we need `default` here since we already have `required` set to True.

Comment (xinyulin) [reply sfputil/main.py]: @mihirpat1 Thanks! I just realized that `default` is unnecessary if `required` is set to True. All defaults were removed in the latest update.

File: sfputil/main.py [hunk 8]
```diff
@@ -1991,11 +1992,62 @@ def loopback(port_name, loopback_mode):
...
+    else:
+        click.echo(f"{port_name}: Failed to connect to STATE_DB")
+        sys.exit(EXIT_FAIL)
+
+    if 'host-side' in loopback_mode:
+        lane_mask = ((1 << host_lane_count) - 1) << ((subport - 1) * host_lane_count)
+    elif 'media-side' in loopback_mode:
+        lane_mask = ((1 << media_lane_count) - 1) << ((subport - 1) * media_lane_count)
+    else:
+        lane_mask = 0
```
Comment (prgeor) [inline sfputil/main.py:2039]: @xinyulin can we write a generic function `get_subport_lane_mask(subport, lane_count)`  which will return the lane mask for the particular `subport`

Comment (xinyulin) [reply sfputil/main.py]: @prgeor Yes, it has been added in the latest commit.

File: doc/Command-Reference.md [hunk 9]
```diff
@@ -3106,19 +3106,19 @@ This command is the standard CMIS diagnostic control used for troubleshooting li
 
 - Usage:
   ```
-  sfputil debug loopback PORT_NAME LOOPBACK_MODE
+  sfputil debug loopback PORT_NAME LOOPBACK_MODE <enable/disable>
 
   Set the loopback mode
```
Comment (mihirpat1) [inline doc/Command-Reference.md:3111]:
```suggestion
  Valid values for LOOPBACK_MODE
```

**Review by mihirpat1 — APPROVED** [review]
@xinyulin Approving with a minor comment

File: doc/Command-Reference.md (hunk 9 above)
Comment (xinyulin) [reply doc/Command-Reference.md]: @mihirpat1  Thanks! It has been revised.

**Review by prgeor — APPROVED** [review]

**Review by mihirpat1 — APPROVED** [review]

**Comment by prgeor** [comment]
@bingwang-ms can you cherry pick to 202405? Its a new CLI for debugging.

**Comment by mihirpat1** [comment]
@yxieca Can you please help to cherry pick this to 202311?
MSFT ADO - 29396307

**Pull request merged**